File: backend/app/models/task.py
import enum
import uuid
from datetime import datetime, timezone

from sqlalchemy import (
    Boolean,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class TaskStatus(str, enum.Enum):
    active = "active"
    paused = "paused"
    completed = "completed"
    overdue = "overdue"
    cancelled = "cancelled"


class TaskPriority(str, enum.Enum):
    low = "low"
    medium = "medium"
    high = "high"
    urgent = "urgent"


class Task(Base):
    __tablename__ = "tasks"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("user.id", ondelete="CASCADE"), nullable=False, index=True
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    deadline: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    status: Mapped[TaskStatus] = mapped_column(
        Enum(TaskStatus), nullable=False, default=TaskStatus.active
    )
    priority: Mapped[TaskPriority] = mapped_column(
        Enum(TaskPriority), nullable=False, default=TaskPriority.medium
    )
    reward_amount: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    penalty_rate: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    is_completed: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    tags: Mapped[list["Tag"]] = relationship(  # noqa: F821
        "Tag", secondary="task_tags", back_populates="tasks"
    )
    work_sessions: Mapped[list["WorkSession"]] = relationship(  # noqa: F821
        "WorkSession", back_populates="task", cascade="all, delete-orphan"
    )

File: frontend/src/types/task.ts
export type TaskStatus = "active" | "paused" | "completed" | "overdue" | "cancelled";
export type TaskPriority = "low" | "medium" | "high" | "urgent";

export interface Tag {
  id: string;
  name: string;
  color: string;
}

export interface Task {
  id: string;
  title: string;
  description: string | null;
  notes: string | null;
  deadline: string;
  status: TaskStatus;
  priority: TaskPriority;
  reward_amount: number;
  penalty_rate: number;
  is_completed: boolean;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
  tags: Tag[];
}

export interface WorkSession {
  id: string;
  task_id: string;
  started_at: string;
  ended_at: string | null;
  duration_seconds: number | null;
  session_type: "manual" | "pomodoro";
}

export interface CurrencyBalance {
  balance: number;
  total_earned: number;
  total_lost: number;
}

export interface Habit {
  id: string;
  name: string;
  color: string;
  cadence: "daily" | "weekly" | "monthly";
  reward_amount: number;
  is_active: boolean;
  created_at: string;
}

export interface HabitEntry {
  id: string;
  habit_id: string;
  date: string;
  completed: boolean;
}

export interface HabitHistoryEntry {
  date: string;
  completed: boolean;
}
